### PR TITLE
[profiler] _RecordFunctionFast - faster python bindings for record_function

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #107720
* #106351
* __->__ #107195

torch.profiler.record_function is relatively slow; for example, in some benchmarks I was running, x.view_as(x) was ~2us, and ~16-17us when wrapped in a record_function context. The reasons for this are: dispatcher overhead from going through an op (the main source of overhead), python binding / python conversion overhead, and some overhead from the context manager.

This new implementation is faster, but it won't work with torchscript. Based on the benchmarks I was running, it adds 0.5-0.7us overhead per call when the profiler is turned off. To use it, you can just:

```python
with torch._C._profiler_manual._RecordFunctionFast("title"):
    torch.add(x, y)
```

It implements a context manager in python which directly calls the record_function utilities, instead of calling through an op.
* The context manager is implemented directly in python because the overhead from calling a python function seems non-negligible
* All the record_function calls, python object conversions are guarded on checks for whether the profiler is enabled or not. It seems like this saves a few hundred nanoseconds.

For more details about the experiments I ran to choose this implementation, see [my record_functions experiments branch](https://github.com/pytorch/pytorch/compare/main...davidberard98:pytorch:record-function-fast-experiments?expand=1).

This also adds a `torch.autograd.profiler._is_profiler_enabled` global variable that can be used to check whether a profiler is currently enabled. It's useful for further reducing the overhead, like this:

```python
if torch.autograd.profiler._is_profiler_enabled:
    with torch._C._profiler_manual._RecordFunctionFast("title"):
        torch.add(x, y)
else:
    torch.add(x, y)
```

On BERT_pytorch (CPU-bound model), if we add a record_function inside CachedAutotuning.run:
* Naive torch.profiler.record_function() is a ~30% slowdown
* Always wrapping with RecordFunctionFast causes a regression of ~2-4%.
* Guarding with an if statement - any regression is within noise

**Selected benchmark results**: these come from a 2.20GHz machine, GPU build but only running CPU ops; running `x.view_as(x)`, with various record_functions applied (with profiling turned off). For more detailed results see "record_functions experiments branch" linked above (those results are on a different machine, but show the same patterns). Note that the results are somewhat noisy, assume 0.05-0.1us variations

```
Baseline:: 1.7825262546539307 us  # Just running x.view_as(x)
profiled_basic:: 13.600390434265137 us  # torch.profiler.record_function(x) + view_as
precompute_manual_cm_rf:: 2.317216396331787 us  # torch._C._profiler_manual._RecordFunctionFast(), if the context is pre-constructed + view_as
guard_manual_cm_rf:: 1.7994389533996582 us  # guard with _is_profiler_enabled + view_as
```

Differential Revision: [D48421198](https://our.internmc.facebook.com/intern/diff/D48421198)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48421198).

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -283,6 +293,7 @@ def __enter__(self):
         if self.entered:
             raise RuntimeError("Profiler context manager is not reentrant")
         _enable_dynamo_cache_lookup_profiler(True)
+        _set_is_profiler_enabled(True)
```
Comment (albanD) [inline torch/autograd/profiler.py:296]: I'm not very familiar with all the entry points to the profiler. Is this really the only place where the profiler can ever be enabled?

File: torch/csrc/profiler/python/init.cpp [hunk 2]
```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
...
 
+/* [NOTE: RecordFunctionFast]
+ * These are an alternate way to call record_function from python.
+ * The typical context manager is slow (~15us on benchmarks in Aug 2023), which
+ * is usually fine for module-level annotations in python, but slow for per-op
+ * annotations. Part of the reason it is slow is because the calls go through
+ * the dispatcher, in order to make the record_function calls work with
+ * torchscript.
+ *
+ * This implementation is less safe; it doesn't work with torchscript.
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:341]: Can you clarify what less safe means here?
Should we simply say that this doesn't go through the dispatcher and thus won't work with any feature relying on the dispatcher (such as TS or compile).

File: torch/csrc/profiler/python/init.cpp [hunk 3]
```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
...
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* RecordFunctionFast_set_name(PyObject* selfGeneric, PyObject* pyName) {
+  HANDLE_TH_ERRORS
+  auto self = (RecordFunctionFast*)selfGeneric;
+  if (self->name) {
+    Py_DECREF(self->name);
+    self->name = nullptr;
+  }
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:432]:
```suggestion
  Py_Clear(self->name);
```

The current code is not safe as self->name is in a bad state between these two lines (keep in mind that decref can deallocate the object and end up running arbitrary code)

File: torch/csrc/profiler/python/init.cpp [hunk 4]
```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
...
+    auto self = (RecordFunctionFast*)selfGeneric;
+    TORCH_INTERNAL_ASSERT(
+        !self->guard,
+        "Trying to enter a new record_function_fast context but the guard is unexpectedly already set");
+    TORCH_INTERNAL_ASSERT(
+        self->name,
+        "Trying to enter a new record_function_fast context but self->name is not set");
+    self->guard =
+        std::make_unique<at::RecordFunction>(at::RecordScope::FUNCTION);
+    THPUtils_checkString(self->name);
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:406]: nit: why not check that on construction / set?

File: torch/csrc/profiler/python/init.cpp [hunk 5]
```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
...
+    TORCH_INTERNAL_ASSERT(
+        self->guard,
+        "Trying to exit an active record_function_fast context but no guard is set");
+    self->guard.reset();
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* RecordFunctionFast_set_name(PyObject* selfGeneric, PyObject* pyName) {
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:426]: Why do you need this? You're not always passing it in the constructor?

File: torch/csrc/profiler/python/init.cpp [hunk 6]
```diff
@@ -326,5 +330,165 @@ void initPythonBindings(PyObject* module) {
...
+ * The typical context manager is slow (~15us on benchmarks in Aug 2023), which
+ * is usually fine for module-level annotations in python, but slow for per-op
+ * annotations. Part of the reason it is slow is because the calls go through
+ * the dispatcher, in order to make the record_function calls work with
+ * torchscript.
+ *
+ * This implementation is less safe; it doesn't work with torchscript.
+ *
+ * It implements a context manager in C++, which reduces the overhead compared
+ * to using a python context manager that calls into C++.
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:344]: I'm not sure to understand what that means? You mean that having the context manager holding to the guard in python would be slower? How much?

**Review by albanD — COMMENTED** [review]
We

File: torch/autograd/profiler.py (hunk 1 above)
Comment (davidberard98) [reply torch/autograd/profiler.py:296]: It's probably better if we put it at all the callsites of `torch._C._autograd.enable_profiler` - I'll change it. I think the difference is that `with emit_nvtx():` and `with emit_itt():` would not be included in the current implementation

File: torch/csrc/profiler/python/init.cpp (hunk 5 above)
Comment (davidberard98) [reply torch/csrc/profiler/python/init.cpp:426]: Good point, will remove it (this was for some other experiments that turned out to not provide any improvement)

File: torch/csrc/profiler/python/init.cpp (hunk 6 above)
Comment (davidberard98) [reply torch/csrc/profiler/python/init.cpp:344]: [https://github.com/pytorch/pytorch/compare/main...davidberard98:pytorch:record-function-fast-experiments?expand=1](perf experiments) - I was seeing 3.59 us for a python context manager that calls into c++ for `enter` and `exit`; vs 3.12 us for the pure-C++ context manager. (results were pretty noisy, so add +/- .1us).

My suspicion was that the python context manager is slower because it needs to perform additional function calls?

**Comment by davidberard98** [comment]
I moved the bindings from `torch._C._profiler_manual._RecordFunctionFast` to `torch._C._profiler._RecordFunctionFast`, following Ed's example from CapturedTraceback

File: test/profiler/test_profiler.py [hunk 7]
```diff
@@ -1688,6 +1688,87 @@ def test_concrete_inputs_profiling_toggling(self):
...
+                    with torch._C._profiler._RecordFunctionFast("add_test_fast_rf5"):
+                        x.relu()
+
+        self.assertGreaterEqual(len([e for e in p.events() if e.name == "add_test_fast_rf4"]), 4)
+        self.assertGreaterEqual(len([e for e in p.events() if e.name == "add_test_fast_rf5"]), 4)
+
+        with profile() as p:
+            cm = torch._C._profiler._RecordFunctionFast(None)
+            for i in range(4):
+                cm.set_name(f"add_test_fast_rf_{i}")
```
Comment (albanD) [inline test/profiler/test_profiler.py:1736]: This can be removed now that we don't have this anymore.

File: torch/csrc/profiler/python/init.cpp [hunk 8]
```diff
@@ -431,6 +433,137 @@ void initPythonBindings(PyObject* module) {
     return py_symbolize(tb_ptrs);
   });
   installCapturedTracebackPython();
+  initRecordFunctionFastBindings(m.ptr());
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:436]: nit: this could really be inlined here tbh. But as you prefer

File: torch/csrc/profiler/python/init.cpp [hunk 9]
```diff
@@ -431,6 +433,137 @@ void initPythonBindings(PyObject* module) {
...
+  auto self = (RecordFunctionFast*)selfGeneric;
+  constexpr const char* kwlist[] = {"name", nullptr};
+  PyObject* name = nullptr;
+  if (!PyArg_ParseTupleAndKeywords(
+          args, kwargs, "O", const_cast<char**>(kwlist), &name)) {
+    return -1;
+  }
+  if (name) {
+    THPUtils_checkString(name);
+    self->name = name;
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:494]: The object returned by PyArg_ParseTupleAndKeywords is a borrowed reference. You need to incref it here and decref it during the dealloc.

File: torch/csrc/profiler/python/init.cpp [hunk 10]
```diff
@@ -431,6 +433,137 @@ void initPythonBindings(PyObject* module) {
...
+      {NULL},
+  };
+
+  static PyTypeObject RecordFunctionFast_Type = {
+      PyVarObject_HEAD_INIT(NULL, 0)};
+
+  RecordFunctionFast_Type.tp_name = "torch._C._profiler.RecordFunctionFast",
+  RecordFunctionFast_Type.tp_basicsize = sizeof(RecordFunctionFast);
+  RecordFunctionFast_Type.tp_dealloc = (destructor)RecordFunctionFast_dealloc;
+  RecordFunctionFast_Type.tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE;
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:549]: do you really need Py_TPFLAGS_BASETYPE here?

**Review by albanD — COMMENTED** [review]
refcount needs fixing. Sounds good otherwise.

File: torch/csrc/profiler/python/init.cpp [hunk 11]
```diff
@@ -112,6 +112,107 @@ struct type_caster<std::shared_ptr<torch::CapturedTraceback>> {
...
+  auto self = (RecordFunctionFast*)selfGeneric;
+  constexpr const char* kwlist[] = {"name", nullptr};
+  PyObject* name = nullptr;
+  if (!PyArg_ParseTupleAndKeywords(
+          args, kwargs, "O", const_cast<char**>(kwlist), &name)) {
+    return -1;
+  }
+  if (name) {
+    Py_INCREF(name);
+    THPUtils_checkString(name);
```
Comment (albanD) [inline torch/csrc/profiler/python/init.cpp:170]:
```suggestion
    TORCH_CHECK(THPUtils_checkString(name), "The name passed to RecordFunction must be a string");
    Py_INCREF(name);
```

The check is never raising and error and we should only incref if no error is raised.

File: torch/profiler/profiler.py [hunk 12]
```diff
@@ -604,13 +604,11 @@ def __init__(
 
     def __enter__(self):
         prof._enable_dynamo_cache_lookup_profiler(True)
-        prof._set_is_profiler_enabled(True)
```
Comment (albanD) [inline torch/profiler/profiler.py]: Why did you remove these?

**Review by albanD — APPROVED** [review]
Sounds good as long as the flag setting removal is intentional.

File: torch/profiler/profiler.py (hunk 12 above)
Comment (davidberard98) [reply torch/profiler/profiler.py]: I moved it to autograd/profiler.py and initially forgot to remove it from this file.

afaik "profiler enabled" == "torch._C._profiler._enable_profiler() has been called", and all the callsites for `_enable_profiler()` are in autograd/profiler.py - so I moved these calls to autograd/profiler.py.

Comment (albanD) [reply torch/profiler/profiler.py]: ok! All good then!

**Review by aaronenyeshi — APPROVED** [review]
This is great! Thank you @davidberard98 for preparing the diff, and performing the benchmark measurements. Also, thank you @albanD for the thorough reviews.
